paddle/fluid/operators/dropout_op.cu

```diff
@@ -13,92 +13,62 @@ See the License for the specific language governing permissions and
 limitations under the License. */
 
 #define EIGEN_USE_GPU
-#include <thrust/device_ptr.h>
-#include <thrust/iterator/counting_iterator.h>
 #include <thrust/random.h>
-#include <thrust/transform.h>
 #include "paddle/fluid/operators/dropout_op.h"
 #include "paddle/fluid/platform/float16.h"
 
 namespace paddle {
 namespace operators {
 
-template <typename T>
-__global__ void RandomGenerator(const size_t n, const int seed,
-                                const float dropout_prob, const T* src,
-                                T* mask_data, T* dst) {
-  thrust::minstd_rand rng;
-  rng.seed(seed);
-  thrust::uniform_real_distribution<float> dist(0, 1);
+template <typename T, typename MaskType>
+__global__ void RandomGenerator(const size_t n, const size_t seed,
+                                const uint32_t int_dropout_prob, const T *src,
+                                MaskType *mask_data, T *dst) {
+  constexpr uint32_t kUInt32Max = static_cast<uint32_t>(-1);
+  size_t idx = blockDim.x * blockIdx.x + threadIdx.x;
 
-  int idx = blockDim.x * blockIdx.x + threadIdx.x;
-  int step_size = 0;
+  thrust::minstd_rand rng;
+  thrust::uniform_int_distribution<uint32_t> dist(0, kUInt32Max - 1);
+  rng.seed(HashCombine(seed, idx));
+  rng.seed(dist(rng));
```

Comment:
If I do not add `rng.seed(dist(rng))`, `nan` may occur in some test cases even though the actual `dropout_num/total_element_num` is still close to `dropout_prob`. The reason may be that there are lots of continuous zeros in `Mask`. Therefore, maybe it is not a best way to produce random seed by `HashCombine`.